qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,608 @@
+
+        Args:
+            host: The hostname from the URL.
+        """
+        return bool(self.dnsResolve(host))
+
+    @_js_slot(str, str, str)
+    def isInNet(self, host, pattern, mask):
+        """Test if host IP is in network.
+
+        Return True if and only if the IP address of the host matches the
+        specified IP address pattern. Pattern and mask specification is done
+        the same way as for SOCKS configuration.
+
+        Args:
+            host: A DNS hostname, or IP address. If a hostname is passed, it
+                  will be resolved into an IP address by this function.
+            pattern: an IP address pattern in the dot-separated format.
+            mask: mask for the IP address pattern informing which parts of
+                  the IP address should be matched against. 0 means ignore,
+                  255 means match.
+        """
+        host_ip = ipaddress.ip_address(host)
+        network = ipaddress.ip_network("{}/{}".format(pattern, mask))
+        return host_ip in network
+
+    @_js_slot(str)
+    def dnsResolve(self, host):
+        """Resolve a DNS hostname.
+
+        Resolves the given DNS hostname into an IP address, and returns it
+        in the dot-separated format as a string.
+
+        Args:
+            host: hostname to resolve.
+        """
+        ips = QHostInfo.fromName(host)
+        if ips.error() != QHostInfo.NoError or len(ips.addresses()) == 0:
+            return ""
+        else:
+            return ips.addresses()[0].toString()
+
+    @_js_slot()
+    def myIpAddress(self):
+        """Get host IP address.
+
+        Return the server IP address of the current machine, as a string in
+        the dot-separated integer format.
+        """
+        return self.dnsResolve(QHostInfo.localHostName())
+
+    @_js_slot(str)
+    def dnsDomainLevels(self, host):
+        """Count number of DNS domain levels.
+
+        Return the number (integer) of DNS domain levels (number of dots)
+        in the hostname.
+
+        Args:
+            host: the hostname from the URL.
+        """
+        return host.count('.')
+
+    @_js_slot(str, str)
+    def shExpMatch(self, mstr, shexp):
+        """Test is string matches shell expression.
+
+        Return true if the string matches the specified shell expression.
+        Currently, the patterns are shell expressions, not regular
+        expressions.
+
+        Args:
+            mstr: Any string to compare (e.g. the URL, or the hostname).
+            shexp: a shell expression to compare against.
+        """
+        return fnmatch.fnmatchcase(mstr, shexp)
+
+    def _get_dt(self, gmt):
+        if self._fixed_dt is not None:
+            if gmt:
+                return self._fixed_dt.toUTC()
+            else:
+                return self._fixed_dt
+        else:
+            if gmt:
+                return QDateTime.currentDateTimeUtc()
+            else:
+                return QDateTime.currentDateTime()
```

Comment:
Can you add a test for this without `_fixed_dt`, just to make this covered and make sure it doesn't crash at least?